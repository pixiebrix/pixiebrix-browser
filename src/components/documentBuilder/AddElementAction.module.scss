.toggle {
  padding: 5px;
}
