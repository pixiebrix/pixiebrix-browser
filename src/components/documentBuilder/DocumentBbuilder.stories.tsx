/*
 * Copyright (C) 2021 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { ComponentMeta, ComponentStory } from "@storybook/react";
import { Formik } from "formik";
import React, { Suspense, useState } from "react";
import * as yup from "yup";
import {
  Button,
  Col,
  Container,
  Form as BootstrapForm,
  Row,
} from "react-bootstrap";
import { action } from "@storybook/addon-actions";
import DocumentEditor from "./DocumentEditor";
import DocumentPreview from "./DocumentPreview";

const schemaShape: yup.ObjectSchema = yup.object().shape({
  body: yup.string().required(),
});

const initialValues = {
  body: `
type: container
children:
  - type: row
    children:
      - type: column
        children:
          - type: header_1
            config:
              title: My document
          - type: header_2
            config:
              title: Testing the doc
              className: text-danger
          - type: text
            config:
              text: Testing the doc`,
};

const DocumentBuilder: React.FC = () => {
  const [activeElement, setActiveElement] = useState<string>(null);

  return (
    <Formik
      validationSchema={schemaShape}
      initialValues={initialValues}
      onSubmit={action("onSubmit")}
    >
      {({ handleSubmit }) => (
        <BootstrapForm noValidate onSubmit={handleSubmit}>
          <Container>
            <Row>
              <Col>
                <DocumentEditor name="body" />
                <Button type="submit">Submit</Button>
              </Col>
              <Col>
                <DocumentPreview name="body" />
              </Col>
            </Row>
          </Container>
        </BootstrapForm>
      )}
    </Formik>
  );
};

const componentMeta: ComponentMeta<typeof DocumentBuilder> = {
  title: "Document Builder",
  component: DocumentBuilder,
};

const DocumentBuilderTemplate: ComponentStory<typeof DocumentBuilder> = () => (
  <DocumentBuilder />
);

export const Default = DocumentBuilderTemplate.bind({});

export default componentMeta;
