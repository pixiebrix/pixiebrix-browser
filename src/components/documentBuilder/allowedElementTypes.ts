/*
 * Copyright (C) 2022 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import {
  DOCUMENT_ELEMENT_TYPES,
  DocumentElement,
  DocumentElementType,
} from "./documentBuilderTypes";

export const ROOT_ELEMENT_TYPES: DocumentElementType[] = [
  "header",
  "text",
  "image",
  "container",
  "card",
  "pipeline",
  "button",
  "list",
];

export const PARENT_ELEMENT_TYPES: DocumentElementType[] = [
  "row",
  "column",
  "container",
  "card",
  "list",
];

const allowedChildTypes: Record<string, DocumentElementType[]> = {
  container: ["row", "list"],
  row: ["column", "list"],
  column: ["header", "text", "image", "card", "pipeline", "button", "list"],
  card: ["header", "text", "image", "container", "pipeline", "button", "list"],
  // Any element we can add to the list
  list: DOCUMENT_ELEMENT_TYPES as unknown as DocumentElementType[],
};

export function getAllowedChildTypes(
  parentElement: DocumentElement
): DocumentElementType[] {
  return allowedChildTypes[parentElement.type] ?? [];
}
