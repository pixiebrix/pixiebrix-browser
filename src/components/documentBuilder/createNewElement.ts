/*
 * Copyright (C) 2021 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { DocumentElement, DocumentElementType } from "./documentBuilderTypes";
import { Expression } from "@/core";
import { BlockConfig, BlockPipeline } from "@/blocks/types";
import { MarkdownRenderer } from "@/blocks/renderers/markdown";
import { uuidv4 } from "@/types/helpers";
import { defaultBlockConfig } from "@/blocks/util";

export function createNewElement(elementType: DocumentElementType) {
  const element: DocumentElement = {
    type: elementType,
    config: {},
  };

  switch (elementType) {
    case "header_1":
    case "header_2":
    case "header_3":
      element.config.title = "Header";
      break;

    case "text":
      element.config.text = "Paragraph text.";
      break;

    case "container":
      element.children = [createNewElement("row")];
      break;

    case "row":
      element.children = [createNewElement("column")];
      break;

    case "column":
      element.children = [];
      break;

    case "card":
      element.config.heading = "Header";
      element.children = [];
      break;

    case "pipeline": {
      const markdownBlock = new MarkdownRenderer();
      const markdownConfig: BlockConfig = {
        id: markdownBlock.id,
        instanceId: uuidv4(),
        config: defaultBlockConfig(markdownBlock.inputSchema),
      };
      element.config.pipeline = {
        __type__: "pipeline",
        __value__: [markdownConfig],
      } as Expression<BlockPipeline, "pipeline">;
      break;
    }

    case "button":
      element.config.title = "Click me";
      break;

    case "list":
      element.config.element = {
        __type__: "defer",
        __value__: createNewElement("text"),
      } as Expression<DocumentElement, "defer">;
      break;

    default:
      throw new Error(
        `Can't create new element. Type "${elementType} is not supported.`
      );
  }

  return element;
}
