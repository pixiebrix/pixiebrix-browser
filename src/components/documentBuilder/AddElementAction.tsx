/*
 * Copyright (C) 2021 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { useField } from "formik";
import React from "react";
import EllipsisMenu from "@/components/ellipsisMenu/EllipsisMenu";
import {
  DocumentElement,
  DocumentElementType,
  DOCUMENT_ELEMENT_TYPES,
} from "./documentBuilderTypes";
import styles from "./AddElementAction.module.scss";
import { createNewElement } from "./createNewElement";

type AddElementActionProps = {
  elementsCollectionName: string;
  allowedTypes?: DocumentElementType[];
  className?: string;
  menuBoundary?: Element;
};

const AddElementAction: React.FC<AddElementActionProps> = ({
  elementsCollectionName,
  allowedTypes = DOCUMENT_ELEMENT_TYPES,
  className,
  menuBoundary,
}) => {
  const [{ value: elementsCollection }, , { setValue }] = useField<
    DocumentElement[]
  >(elementsCollectionName);

  const addElement = (elementType: DocumentElementType) => {
    const element = createNewElement(elementType);
    setValue([...elementsCollection, element]);
  };

  return (
    <EllipsisMenu
      className={className}
      toggleClassName={styles.toggle}
      items={allowedTypes.map((elementType) => ({
        title: elementType,
        action: () => {
          addElement(elementType);
        },
      }))}
      menuBoundary={menuBoundary}
    />
  );
};

export default AddElementAction;
