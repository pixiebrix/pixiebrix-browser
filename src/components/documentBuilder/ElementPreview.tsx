/*
 * Copyright (C) 2021 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React, { MouseEventHandler } from "react";
import styles from "./ElementPreview.module.scss";
import cx from "classnames";
import { DocumentElement } from "./documentBuilderTypes";
import { getPreviewComponentDefinition } from "./documentTree";
import AddElementAction from "./AddElementAction";
import { useField } from "formik";
import { getAllowedChildTypes } from "./allowedElementTypes";

interface ElementPreviewTemplateProps {
  elementName: string;
  activeElement: string;
  setActiveElement: (name: string) => void;
  menuBoundary?: Element;
}

const ElementPreview: React.FC<ElementPreviewTemplateProps> = ({
  elementName,
  activeElement,
  setActiveElement,
  menuBoundary,
}) => {
  const [{ value: documentElement }] = useField<DocumentElement>(elementName);
  const isActive = activeElement === elementName;
  const onClick: MouseEventHandler<HTMLDivElement> = (event) => {
    event.stopPropagation();

    if (!isActive) {
      setActiveElement(elementName);
    }
  };

  const isContainer = Array.isArray(documentElement.children);

  const { Component: PreviewComponent, props } = getPreviewComponentDefinition(
    documentElement
  );

  return (
    <PreviewComponent
      {...props}
      onClick={onClick}
      className={cx(props?.className, styles.root, {
        [styles.active]: isActive,
      })}
    >
      {props?.children}
      {isContainer &&
        documentElement.children.map((childElement, i) => (
          <ElementPreview
            key={`${elementName}.children.${i}`}
            elementName={`${elementName}.children.${i}`}
            activeElement={activeElement}
            setActiveElement={setActiveElement}
            menuBoundary={menuBoundary}
          />
        ))}
      {isContainer && (
        <AddElementAction
          as="ellipsis"
          elementsCollectionName={`${elementName}.children`}
          allowedTypes={getAllowedChildTypes(documentElement)}
          className={styles.addElement}
          menuBoundary={menuBoundary}
        />
      )}
    </PreviewComponent>
  );
};

export default ElementPreview;
