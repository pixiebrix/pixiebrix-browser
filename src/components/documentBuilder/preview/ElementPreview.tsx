/*
 * Copyright (C) 2021 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React, { MouseEventHandler, useMemo } from "react";
import styles from "./ElementPreview.module.scss";
import cx from "classnames";
import {
  DocumentElement,
  isListElement,
} from "@/components/documentBuilder/documentBuilderTypes";
import { getPreviewComponentDefinition } from "@/components/documentBuilder/documentTree";
import AddElementAction from "./AddElementAction";
import { useField } from "formik";
import { getAllowedChildTypes } from "@/components/documentBuilder/allowedElementTypes";

export type ElementPreviewProps = {
  elementName: string;
  activeElement: string | null;
  setActiveElement: (name: string | null) => void;
  hoveredElement: string | null;
  setHoveredElement: (name: string | null) => void;
  menuBoundary?: Element;
};

const ElementPreview: React.FC<ElementPreviewProps> = ({
  elementName,
  activeElement,
  setActiveElement,
  hoveredElement,
  setHoveredElement,
  menuBoundary,
}) => {
  const [{ value: documentElement }] = useField<DocumentElement>(elementName);
  const isActive = activeElement === elementName;
  const isHovered = hoveredElement === elementName && !isActive;
  const onClick: MouseEventHandler<HTMLDivElement> = (event) => {
    event.stopPropagation();

    if (!isActive) {
      setActiveElement(elementName);
    }
  };

  const onMouseOver: MouseEventHandler<HTMLDivElement> = (event) => {
    event.stopPropagation();
    if (hoveredElement !== elementName) {
      setHoveredElement(elementName);
    }
  };

  const onMouseLeave: MouseEventHandler<HTMLDivElement> = () => {
    if (hoveredElement === elementName) {
      setHoveredElement(null);
    }
  };

  // Render children and Add Menu for the container element
  const isContainer = Array.isArray(documentElement.children);

  // Render the item template and the Item Type Selector for the list element
  const isList = isListElement(documentElement);

  const { Component: PreviewComponent, props } = useMemo(
    () => getPreviewComponentDefinition(documentElement),
    [documentElement]
  );

  return (
    <PreviewComponent
      {...props}
      onClick={onClick}
      className={cx(props?.className, styles.root, {
        [styles.active]: isActive,
        [styles.hovered]: isHovered,
      })}
      onMouseOver={onMouseOver}
      onMouseLeave={onMouseLeave}
    >
      {props?.children}
      {isContainer &&
        documentElement.children.map((childElement, i) => (
          <ElementPreview
            key={`${elementName}.children.${i}`}
            elementName={`${elementName}.children.${i}`}
            activeElement={activeElement}
            setActiveElement={setActiveElement}
            menuBoundary={menuBoundary}
            hoveredElement={hoveredElement}
            setHoveredElement={setHoveredElement}
          />
        ))}
      {isContainer && (
        <AddElementAction
          elementsCollectionName={`${elementName}.children`}
          allowedTypes={getAllowedChildTypes(documentElement)}
          className={styles.addElement}
          menuBoundary={menuBoundary}
        />
      )}
      {isList && (
        <ElementPreview
          elementName={`${elementName}.config.element.__value__`}
          activeElement={activeElement}
          setActiveElement={setActiveElement}
          menuBoundary={menuBoundary}
          hoveredElement={hoveredElement}
          setHoveredElement={setHoveredElement}
        />
      )}
    </PreviewComponent>
  );
};

export default ElementPreview;
