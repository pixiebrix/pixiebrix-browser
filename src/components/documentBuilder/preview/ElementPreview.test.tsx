/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React from "react";
import {
  DOCUMENT_ELEMENT_TYPES,
  type DocumentElement,
  type DocumentElementType,
} from "@/components/documentBuilder/documentBuilderTypes";
import { createNewElement } from "@/components/documentBuilder/createNewElement";
import ElementPreview, {
  type ElementPreviewProps,
} from "@/components/documentBuilder/preview/ElementPreview";
import { fireEvent } from "@testing-library/react";
import { act } from "react-dom/test-utils";
import { defaultBlockConfig } from "@/bricks/util";
import { MarkdownRenderer } from "@/bricks/renderers/markdown";
import { type PipelineExpression } from "@/types/runtimeTypes";
import { render } from "@/pageEditor/testHelpers";
import { actions } from "@/pageEditor/slices/editorSlice";
import userEvent from "@testing-library/user-event";
import {
  baseExtensionStateFactory,
  formStateFactory,
} from "@/testUtils/factories/pageEditorFactories";
import { brickConfigFactory } from "@/testUtils/factories/brickFactories";

window.HTMLElement.prototype.scrollIntoView = jest.fn();

const renderElementPreview = (
  element: DocumentElement,
  elementPreviewProps?: Partial<ElementPreviewProps>
) => {
  const props: ElementPreviewProps = {
    documentBodyName: "",
    elementName: "element",
    previewElement: element,
    activeElement: null,
    setActiveElement: jest.fn(),
    hoveredElement: null,
    setHoveredElement: jest.fn(),
    ...elementPreviewProps,
  };

  const formState = formStateFactory({
    extension: baseExtensionStateFactory({
      blockPipeline: [
        brickConfigFactory({
          config: {
            body: [element],
          },
        }),
      ],
    }),
  });

  return render(<ElementPreview {...props} />, {
    initialValues: formState,
    setupRedux(dispatch) {
      dispatch(actions.addElement(formState));
      dispatch(actions.selectElement(formState.uuid));
      dispatch(
        actions.setElementActiveNodeId(
          formState.extension.blockPipeline[0].instanceId
        )
      );
      dispatch(actions.setNodePreviewActiveElement("0"));
    },
  });
};

test("calls setActiveElement callback on click", async () => {
  const setActiveElementMock = jest.fn();
  const element = createNewElement("text");
  const { container } = renderElementPreview(element, {
    setActiveElement: setActiveElementMock,
  });

  await act(async () => {
    fireEvent.click(container.querySelector("p"));
  });
  expect(setActiveElementMock).toHaveBeenCalledWith("element");
});

test("prevents navigation on link click", async () => {
  const consoleError = jest.spyOn(global.console, "error");

  const element = createNewElement("text");
  element.config.text = "Link in markdown [www.google.com](https://google.com)";
  const rendered = renderElementPreview(element, {
    setActiveElement: jest.fn(),
  });

  await userEvent.click(rendered.getByText("www.google.com"));
  expect(rendered.getByText("Link in markdown")).toBeInTheDocument();

  // On navigation jest calls console.error, fail the test if it is called
  expect(consoleError).not.toHaveBeenCalled();

  consoleError.mockRestore();
});

test("adds a CSS class to an active element", async () => {
  const element = createNewElement("text");
  const { container } = renderElementPreview(element, {
    activeElement: "element",
  });

  expect(container.querySelector("div")).toHaveClass("active");
  expect(window.HTMLElement.prototype.scrollIntoView).toHaveBeenCalled();
});

test("calls setHoveredElement callback on hover", async () => {
  const setHoveredElementMock = jest.fn();
  const element = createNewElement("text");
  const { container } = renderElementPreview(element, {
    setHoveredElement: setHoveredElementMock,
  });

  await act(async () => {
    fireEvent.mouseOver(container.querySelector("p"));
  });
  expect(setHoveredElementMock).toHaveBeenCalledWith("element");
  setHoveredElementMock.mockClear();
});

test("calls setHoveredElement callback when mouse leaves", async () => {
  const setHoveredElementMock = jest.fn();
  const element = createNewElement("text");
  const { container } = renderElementPreview(element, {
    hoveredElement: "element",
    setHoveredElement: setHoveredElementMock,
  });

  await act(async () => {
    fireEvent.mouseLeave(container.querySelector("p"));
  });
  expect(setHoveredElementMock).toHaveBeenCalledWith(null);
});

test("adds a CSS class to a hovered element", async () => {
  const element = createNewElement("text");
  const { container } = renderElementPreview(element, {
    hoveredElement: "element",
  });

  expect(container.querySelector("div")).toHaveClass("hovered");
});

test.each(
  DOCUMENT_ELEMENT_TYPES.filter(
    (x) => !["header_1", "header_2", "header_3"].includes(x)
  )
)("can preview default %s", (elementType: DocumentElementType) => {
  const element = createNewElement(elementType);
  const rendered = renderElementPreview(element);
  expect(rendered.asFragment()).toMatchSnapshot();
});

test("can preview pipeline element with bricks", () => {
  const testBrick = brickConfigFactory();
  const markdownBlock = new MarkdownRenderer();
  const markdownConfig = brickConfigFactory({
    id: markdownBlock.id,
    config: defaultBlockConfig(markdownBlock.inputSchema),
  });

  const element = createNewElement("pipeline");
  const pipeline = element.config.pipeline as PipelineExpression;
  pipeline.__value__.push(testBrick, markdownConfig);
  const rendered = renderElementPreview(element);

  expect(rendered.asFragment()).toMatchSnapshot();
});
