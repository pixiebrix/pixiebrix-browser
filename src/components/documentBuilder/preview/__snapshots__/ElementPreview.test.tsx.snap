// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can preview default button 1`] = `
<DocumentFragment>
  <form
    action="#"
  >
    <div>
      <div
        class="root inlineWrapper"
      >
        <button
          class="btn btn-primary"
          type="button"
        >
          Action
        </button>
      </div>
    </div>
    <button
      type="submit"
    >
      Submit
    </button>
  </form>
</DocumentFragment>
`;

exports[`can preview default card 1`] = `
<DocumentFragment>
  <form
    action="#"
  >
    <div
      class="shiftRightWrapper root"
    >
      <div
        class="card"
      >
        <div
          class="card-header"
        >
          Header
        </div>
        <div
          class="container cardBody overflow-auto card-body"
        >
          <div
            class="addElement dropdown"
          >
            <button
              aria-expanded="false"
              aria-haspopup="true"
              class="toggle toggle dropdown-toggle btn btn-light btn-sm"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-ellipsis-v fa-w-6 "
                data-icon="ellipsis-v"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 192 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M96 184c39.8 0 72 32.2 72 72s-32.2 72-72 72-72-32.2-72-72 32.2-72 72-72zM24 80c0 39.8 32.2 72 72 72s72-32.2 72-72S135.8 8 96 8 24 40.2 24 80zm0 352c0 39.8 32.2 72 72 72s72-32.2 72-72-32.2-72-72-72-72 32.2-72 72z"
                  fill="currentColor"
                />
              </svg>
            </button>
          </div>
        </div>
      </div>
    </div>
    <button
      type="submit"
    >
      Submit
    </button>
  </form>
</DocumentFragment>
`;

exports[`can preview default column 1`] = `
<DocumentFragment>
  <form
    action="#"
  >
    <div
      class="root container col"
    >
      <span
        class="text-muted"
      >
        Column
      </span>
      <div
        class="addElement dropdown"
      >
        <button
          aria-expanded="false"
          aria-haspopup="true"
          class="toggle toggle dropdown-toggle btn btn-light btn-sm"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-ellipsis-v fa-w-6 "
            data-icon="ellipsis-v"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 192 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M96 184c39.8 0 72 32.2 72 72s-32.2 72-72 72-72-32.2-72-72 32.2-72 72-72zM24 80c0 39.8 32.2 72 72 72s72-32.2 72-72S135.8 8 96 8 24 40.2 24 80zm0 352c0 39.8 32.2 72 72 72s72-32.2 72-72-32.2-72-72-72-72 32.2-72 72z"
              fill="currentColor"
            />
          </svg>
        </button>
      </div>
    </div>
    <button
      type="submit"
    >
      Submit
    </button>
  </form>
</DocumentFragment>
`;

exports[`can preview default container 1`] = `
<DocumentFragment>
  <form
    action="#"
  >
    <div
      class="root container container"
    >
      <div
        class="root container row"
      >
        <div
          class="root container col"
        >
          <span
            class="text-muted"
          >
            Column
          </span>
          <div
            class="addElement dropdown"
          >
            <button
              aria-expanded="false"
              aria-haspopup="true"
              class="toggle toggle dropdown-toggle btn btn-light btn-sm"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-ellipsis-v fa-w-6 "
                data-icon="ellipsis-v"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 192 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M96 184c39.8 0 72 32.2 72 72s-32.2 72-72 72-72-32.2-72-72 32.2-72 72-72zM24 80c0 39.8 32.2 72 72 72s72-32.2 72-72S135.8 8 96 8 24 40.2 24 80zm0 352c0 39.8 32.2 72 72 72s72-32.2 72-72-32.2-72-72-72-72 32.2-72 72z"
                  fill="currentColor"
                />
              </svg>
            </button>
          </div>
        </div>
        <div
          class="addElement dropdown"
        >
          <button
            aria-expanded="false"
            aria-haspopup="true"
            class="toggle toggle dropdown-toggle btn btn-light btn-sm"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-ellipsis-v fa-w-6 "
              data-icon="ellipsis-v"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 192 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M96 184c39.8 0 72 32.2 72 72s-32.2 72-72 72-72-32.2-72-72 32.2-72 72-72zM24 80c0 39.8 32.2 72 72 72s72-32.2 72-72S135.8 8 96 8 24 40.2 24 80zm0 352c0 39.8 32.2 72 72 72s72-32.2 72-72-32.2-72-72-72-72 32.2-72 72z"
                fill="currentColor"
              />
            </svg>
          </button>
        </div>
      </div>
      <div
        class="addElement dropdown"
      >
        <button
          aria-expanded="false"
          aria-haspopup="true"
          class="toggle toggle dropdown-toggle btn btn-light btn-sm"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-ellipsis-v fa-w-6 "
            data-icon="ellipsis-v"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 192 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M96 184c39.8 0 72 32.2 72 72s-32.2 72-72 72-72-32.2-72-72 32.2-72 72-72zM24 80c0 39.8 32.2 72 72 72s72-32.2 72-72S135.8 8 96 8 24 40.2 24 80zm0 352c0 39.8 32.2 72 72 72s72-32.2 72-72-32.2-72-72-72-72 32.2-72 72z"
              fill="currentColor"
            />
          </svg>
        </button>
      </div>
    </div>
    <button
      type="submit"
    >
      Submit
    </button>
  </form>
</DocumentFragment>
`;

exports[`can preview default header 1`] = `
<DocumentFragment>
  <form
    action="#"
  >
    <div
      class="shiftRightWrapper root"
    >
      <h1>
        Header
      </h1>
    </div>
    <button
      type="submit"
    >
      Submit
    </button>
  </form>
</DocumentFragment>
`;

exports[`can preview default image 1`] = `
<DocumentFragment>
  <form
    action="#"
  >
    <div
      class="imageWrapper root"
    >
      <img
        alt="Placeholder"
        class=""
        src="holder.js/100x50"
      />
    </div>
    <button
      type="submit"
    >
      Submit
    </button>
  </form>
</DocumentFragment>
`;

exports[`can preview default list 1`] = `
<DocumentFragment>
  <form
    action="#"
  >
    <div
      class="root container listContainer"
    >
      <div
        class="text-muted"
      >
        List: undefined
      </div>
      <div
        class="text-muted"
      >
        Element key: @element
      </div>
      <div
        class="shiftRightWrapper root"
      >
        <div>
          <p>
            Paragraph text. 
            <strong>
              Markdown
            </strong>
             is supported.
          </p>
          

        </div>
      </div>
    </div>
    <button
      type="submit"
    >
      Submit
    </button>
  </form>
</DocumentFragment>
`;

exports[`can preview default pipeline 1`] = `
<DocumentFragment>
  <form
    action="#"
  >
    <div
      class="shiftRightWrapper root"
    >
      <h3>
        Brick
      </h3>
    </div>
    <button
      type="submit"
    >
      Submit
    </button>
  </form>
</DocumentFragment>
`;

exports[`can preview default row 1`] = `
<DocumentFragment>
  <form
    action="#"
  >
    <div
      class="root container row"
    >
      <div
        class="root container col"
      >
        <span
          class="text-muted"
        >
          Column
        </span>
        <div
          class="addElement dropdown"
        >
          <button
            aria-expanded="false"
            aria-haspopup="true"
            class="toggle toggle dropdown-toggle btn btn-light btn-sm"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-ellipsis-v fa-w-6 "
              data-icon="ellipsis-v"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 192 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M96 184c39.8 0 72 32.2 72 72s-32.2 72-72 72-72-32.2-72-72 32.2-72 72-72zM24 80c0 39.8 32.2 72 72 72s72-32.2 72-72S135.8 8 96 8 24 40.2 24 80zm0 352c0 39.8 32.2 72 72 72s72-32.2 72-72-32.2-72-72-72-72 32.2-72 72z"
                fill="currentColor"
              />
            </svg>
          </button>
        </div>
      </div>
      <div
        class="addElement dropdown"
      >
        <button
          aria-expanded="false"
          aria-haspopup="true"
          class="toggle toggle dropdown-toggle btn btn-light btn-sm"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-ellipsis-v fa-w-6 "
            data-icon="ellipsis-v"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 192 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M96 184c39.8 0 72 32.2 72 72s-32.2 72-72 72-72-32.2-72-72 32.2-72 72-72zM24 80c0 39.8 32.2 72 72 72s72-32.2 72-72S135.8 8 96 8 24 40.2 24 80zm0 352c0 39.8 32.2 72 72 72s72-32.2 72-72-32.2-72-72-72-72 32.2-72 72z"
              fill="currentColor"
            />
          </svg>
        </button>
      </div>
    </div>
    <button
      type="submit"
    >
      Submit
    </button>
  </form>
</DocumentFragment>
`;

exports[`can preview default text 1`] = `
<DocumentFragment>
  <form
    action="#"
  >
    <div
      class="shiftRightWrapper root"
    >
      <div>
        <p>
          Paragraph text. 
          <strong>
            Markdown
          </strong>
           is supported.
        </p>
        

      </div>
    </div>
    <button
      type="submit"
    >
      Submit
    </button>
  </form>
</DocumentFragment>
`;

exports[`can preview pipeline element with bricks 1`] = `
<DocumentFragment>
  <form
    action="#"
  >
    <div
      class="shiftRightWrapper root"
    >
      <h3>
        Brick
      </h3>
      <p>
        testing/block-id_17
      </p>
      <p>
        @pixiebrix/markdown
      </p>
    </div>
    <button
      type="submit"
    >
      Submit
    </button>
  </form>
</DocumentFragment>
`;
